bullet/src/FreeGroupFeatures.cc

```diff
@@ -0,0 +1,82 @@
+#include "FreeGroupFeatures.hh"
+
+namespace ignition {
+namespace physics {
+namespace bullet {
+
+/////////////////////////////////////////////////
+Identity FreeGroupFeatures::FindFreeGroupForModel(
+    const Identity &_modelID) const
+{
+  const auto &model = this->models.at(_modelID)->model;
+
+  // If there are no links at all in this model, then the FreeGroup functions
+  // will not work properly, so we'll just reject these cases.
+  if (model->getNumLinks() == 0)
+    return this->GenerateInvalidId();
+
+  // Reject also if the model has fixed base
+  if (model->hasFixedBase())
+    return this->GenerateInvalidId();
+
+  return _modelID;
+}
+
+/////////////////////////////////////////////////
+Identity FreeGroupFeatures::FindFreeGroupForLink(
+    const Identity &_linkID) const
+{
+  const auto &link_it = this->links.find(_linkID);
+
+  if (link_it != this->links.end() && link_it->second != nullptr)
+    return this->GenerateIdentity(_linkID.id, link_it->second);
```

Comment:
In this case, isn't it the same to just return the `link_it` object?

Reply:
link_it is an iterator, but if you mean returning the same ID, i think FindFreeGroup returns a new link, like in dart https://github.com/ignitionrobotics/ign-physics/blob/main/dartsim/src/FreeGroupFeatures.cc#L85